storage/rocksdb/clone/donor.cc

```diff
@@ -0,0 +1,1462 @@
+      MyOsError(errn, EE_STAT, MYF(0), donor_file_path.c_str());
+      save_error(ER_CANT_OPEN_FILE, donor_file_path, errn);
+      mysql_file_close(fd, MYF(MY_WME));
+      return myrocks::clone::return_error(ER_CANT_OPEN_FILE,
+                                          "Failed to fcntl(F_NOCACHE) a file");
+    }
+#endif
+
+    const auto new_pos =
+        my_seek(fd, metadata.get_sent_offset(), SEEK_SET, MYF(MY_WME));
+    if (new_pos != metadata.get_sent_offset()) {
+      assert(new_pos == MY_FILEPOS_ERROR);
+      mysql_file_close(fd, MYF(MY_WME));
+      save_error(ER_FSEEK_FAIL, donor_file_path, my_errno());
+      return myrocks::clone::return_error(ER_FSEEK_FAIL,
+                                          "Failed to seek a file");
+    }
+
+    if (!metadata.started()) {
+      err = send_metadata(metadata, cbk);
+      if (err != 0) return err;
+    }
+
+    Ha_clone_file clone_file{
+#ifndef __APPLE__
+        use_direct_io,
+#else
+        false,
+#endif
+        Ha_clone_file::FILE_DESC,
+        {fd}};
+    cbk.set_source_name(metadata.get_name().c_str());
+
+    auto buf_size = cbk.get_client_buffer_size();
+    const auto remote_clone = buf_size == 0;
+
+    cbk.clear_flags();
+    if (!use_direct_io) cbk.set_os_buffer_cache();
+    if (!remote_clone) cbk.set_zero_copy();
+
+    if (remote_clone) {
+      // 1M value matches InnoDB clone behavior. We could also consider deriving
+      // the value from i.e. max_allowed_packet value.
+      buf_size = 1024 * 1024;
+    }
+    assert(buf_size > 0);
+
+    while (!metadata.eof()) {
+      err = handle_any_error(thd);
+      if (err != 0) {
+        mysql_file_close(fd, MYF(MY_WME));
+        return err;
+      }
+
+      const auto chunk_size = metadata.chunk_size(buf_size);
+      donor_chunk_metadata chunk{metadata.get_id(), chunk_size};
+      myrocks::Rdb_string_writer buf;
+      chunk.serialize(buf);
+      cbk.set_data_desc(buf.ptr(), buf.get_current_pos());
+
+      DBUG_EXECUTE_IF("myrocks_clone_donor_copy_file_crash", DBUG_SUICIDE(););
+
+      err = cbk.file_cbk(clone_file, chunk_size);
+      if (err != 0) {
+        save_error(err, donor_file_path, my_errno());
+        mysql_file_close(fd, MYF(MY_WME));
+        return myrocks::clone::return_error(err, "Failed to send file data");
+      }
+      metadata.advance(chunk_size);
+    }
+
+    mysql_mutex_lock(&m_donor_mutex);
+    auto in_progress_files_itr = m_in_progress_files.find(metadata.get_id());
+    move_file_metadata(m_in_progress_files, m_completed_files,
+                       std::move(in_progress_files_itr));
+    mysql_mutex_unlock(&m_donor_mutex);
+
+    err = mysql_file_close(fd, MYF(MY_WME));
+    if (err != 0) {
+      save_error(err, donor_file_path, my_errno());
+      return myrocks::clone::return_error(err, "Failed to close a file");
+    }
+
+    err = handle_any_error(thd);
+    if (err != 0) return err;
+
+    mysql_mutex_lock(&m_donor_mutex);
+    if (should_roll_checkpoint()) {
```

Comment:
Holding the lock to create the checkpoint seems unnecessary and would block worker threads from selecting files for transfer. Maybe only task_id 0 should be allowed to create rolling checkpoints? Then you would only need to hold the lock to add the files to the map.

Reply:
I had an implementation where only the master thread could roll the checkpoint, but basically was forced to switch to the one where any thread could do that. The problem was that the master thread could be the very last one to arrive to the InnoDB phase transition, and the stalls can be significant, since there is network I/O involved. In this case, all the workers will be idle instead of precopying for significant duration too.

Ideally rolling the checkpoint should not prevent other threads from continuing work with the older checkpoint, but this is not without issues: 1) rolling already allows the threads with files in-progress to complete them, this limits any potential gains to the next item race winners; 2) rolling deletes the old checkpoint first, so a worker races with the delete to open an SST. If the worker wins, fine, it will send that file. If the worker loses and gets ENOENT, that file is assumed to be an outdated file and is removed from the clone set. The rolled checkpoint would re-add it, but the data size estimates will be thrown off.

I guess we could create a new checkpoint before deleting the old one, but need to think about this more